pkg/localstore/sampler.go

```diff
@@ -0,0 +1,163 @@
+		int64(s.GetDuration.Load()/1000000),
+		int64(s.HmacrDuration.Load()/1000000),
+	)
+}
+
+func (db *DB) ReserveSample(ctx context.Context, anchor []byte, storageDepth uint8) (storage.Sample, error) {
+
+	g, ctx := errgroup.WithContext(ctx)
+	addrChan := make(chan swarm.Address)
+	var stat sampleStat
+	logger := db.logger.WithName("sampler").V(1).Register()
+
+	g.Go(func() error {
+		defer close(addrChan)
+		iterationStart := time.Now()
+		for bin := storageDepth; bin <= swarm.MaxPO; bin++ {
+			err := db.pullIndex.Iterate(func(item shed.Item) (bool, error) {
+				select {
+				case addrChan <- swarm.NewAddress(item.Address):
+					stat.TotalIterated.Inc()
+					return false, nil
+				case <-ctx.Done():
+					return true, ctx.Err()
+				case <-db.close:
+					return true, errors.New("sampler: db closed")
+				}
+			}, &shed.IterateOptions{
+				Prefix: []byte{bin},
+			})
+			if err != nil {
+				logger.Error(err, "sampler: failed iteration")
+				return err
+			}
+		}
+		stat.IterationDuration.Add(time.Since(iterationStart).Microseconds())
+		return nil
+	})
+
+	sampleChan := make(chan storage.SampleItem)
+	const workers = 4
+	for i := 0; i < workers; i++ {
+		g.Go(func() error {
+			hmacr := hmac.New(swarm.NewHasher, anchor)
+
+			for addr := range addrChan {
+				getStart := time.Now()
+				chItem, err := db.get(ctx, storage.ModeGetSync, addr)
+				stat.GetDuration.Add(time.Since(getStart).Microseconds())
+				if err != nil {
+					stat.NotFound.Inc()
+					continue
+				}
+
+				hmacrStart := time.Now()
+				hmacr.Write(chItem.Data)
+				taddr := hmacr.Sum(nil)
+				hmacr.Reset()
+				stat.HmacrDuration.Add(time.Since(hmacrStart).Microseconds())
+
+				select {
+				case sampleChan <- storage.SampleItem{
+					Address:            addr,
+					TransformedAddress: swarm.NewAddress(taddr),
+				}:
+					// continue
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-db.close:
+					return errors.New("sampler: db closed")
+				}
+			}
+
+			return nil
+		})
+	}
+
+	go func() {
+		g.Wait()
+		close(sampleChan)
+	}()
+
+	samples := make([]storage.SampleItem, 0, sampleSize)
+	insert := func(item storage.SampleItem) {
+		added := false
+		for i, sItem := range samples {
+			if leq(item.TransformedAddress.Bytes(), sItem.TransformedAddress.Bytes()) {
+				samples = append(samples[:i+1], samples[i:]...)
+				samples[i] = item
+				added = true
+				break
+			}
+		}
+		if len(samples) > sampleSize {
+			samples = samples[:sampleSize]
+		}
+		if len(samples) < sampleSize && !added {
+			samples = append(samples, item)
+		}
+	}
+
+	for sample := range sampleChan {
+		lowestAddr := swarm.NewAddress(make([]byte, 32))
+		if len(samples) > 0 {
+			lowestAddr = samples[len(samples)-1].TransformedAddress
+		}
+		if leq(sample.TransformedAddress.Bytes(), lowestAddr.Bytes()) || len(samples) < sampleSize {
+			insert(sample)
+		}
+	}
+
+	hasher := swarm.NewHasher()
+	defer hasher.Reset()
+
+	for _, s := range samples {
```

Comment:
here we should be making the assumption that samplesize <=128

Reply:
So the insert function guarantees that we will never have samples exceed the sampleSize. It will pop any item beyond it. So I guess we dont need to check this here unless you think you want to have this extra safety.

Also, currently the sample size is 16. Not sure if this is the agreed number, but during our discussion with abel, we decided this no.